plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunction.java

```diff
@@ -0,0 +1,170 @@
+        checkFunctionArgument(sinceVersion >= 0, "since_version value for function table_changes() cannot be negative");
+
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+        DeltaLakeMetadata deltaLakeMetadata = deltaLakeMetadataFactory.create(session.getIdentity());
+        DeltaLakeTableHandle tableHandle = deltaLakeMetadata.getTableHandle(session, schemaTableName);
+        validateThatCdfWasEnabled(session, fileSystemFactory, tableHandle, sinceVersion);
+        List<DeltaLakeColumnHandle> columnHandles = deltaLakeMetadata.getColumnHandles(session, tableHandle)
+                .values().stream()
+                .map(columnHandle -> (DeltaLakeColumnHandle) columnHandle)
+                .filter(column -> !IGNORED_COLUMNS.contains(column.getName()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<Descriptor.Field> outputFields = ImmutableList.builder();
+        columnHandles.stream()
+                .map(columnHandle -> new Descriptor.Field(columnHandle.getName(), Optional.of(columnHandle.getType())))
+                .forEach(outputFields::add);
+
+        outputFields.add(new Descriptor.Field(CHANGE_TYPE_COLUMN_NAME, Optional.of(VARCHAR)));
+        outputFields.add(new Descriptor.Field(COMMIT_VERSION_COLUMN_NAME, Optional.of(INTEGER)));
+        outputFields.add(new Descriptor.Field(COMMIT_TIMESTAMP_COLUMN_NAME, Optional.of(TIMESTAMP_TZ_MILLIS)));
+
+        return TableFunctionAnalysis.builder()
+                .handle(new TableChangesFunctionTableHandle(schemaName, tableName, sinceVersion, columnHandles))
+                .returnedType(new Descriptor(outputFields.build()))
+                .build();
+    }
+
+    public void validateThatCdfWasEnabled(
+            ConnectorSession session,
+            TrinoFileSystemFactory fileSystemFactory,
+            DeltaLakeTableHandle tableHandle,
+            long sinceVersion)
+    {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        Path transactionLogDir = getTransactionLogDir(new Path(tableHandle.getLocation()));
+        long currentVersion = sinceVersion;
+
+        try {
+            while (currentVersion < tableHandle.getReadVersion()) {
+                Optional<List<DeltaLakeTransactionLogEntry>> entries = getEntriesFromJson(currentVersion, transactionLogDir, fileSystem);
+                if (entries.isPresent()) {
+                    boolean containsRemoveEntry = false;
+                    boolean containsCdcEntry = false;
+                    for (DeltaLakeTransactionLogEntry entry : entries.get()) {
+                        if (entry.getCDC() != null) {
+                            containsCdcEntry = true;
+                        }
+                        if (entry.getRemove() != null) {
+                            containsRemoveEntry = true;
+                        }
+                    }
+                    if (containsRemoveEntry && !containsCdcEntry) {
```

Comment:
Side question: From a user perspective, when meeting this exception, how would I as a user be able to find out where CDF has been enabled?

Reply:
You wouldn't but I think that is fine because if there are only inserts we allow for reading CDF even if it is not enabled for those version so in a scenario when you do the following:
1. create table without cdf
2. insert
3. insert 
4. update
5. insert
and then calls table_changes for versions 2-5 it would fail for versions 4 and in fact it was not enabled for version 2 and 3 so saying something like: CDF is not enabled for version 4  would be weird 
WDYT ?

Reply:
I think that given the fact that we don't explicitly check whether CDF has been enabled, in your example, before or at the version `2` of the table  - the error message currently used is actually fine.
Thank you for explaining it.